### PR TITLE
PERF-#4870: Avoid index materialization in `__getattribute__` and `__getitem__`

### LINKED ISSUE #4870: PERF: try to avoid index materialization in `BasePandasDataset.__getattribute__`
Place: https://github.com/modin-project/modin/blob/cfcdf081e197d35e1fd1526f5585e361793dc2fa/modin/pandas/base.py#L3415

For optimization purpose we use `self.empty` check that trigger index materialization. It is most likely possible to do this check only when the index is already materialized. To do this, we can check the corresponding cache.

### PR DESCRIPTION
**Pull request by anmyachev**
Signed-off-by: Myachev <anatoly.myachev@intel.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #4870 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date 
- [x] added (Issue Number: PR title (PR Number)) and github username to release notes for next major release

### CODE REVIEW

**Comment by anmyachev** [comment]
@vnlitvinov could you review this PR?

File: modin/core/storage_formats/base/query_compiler.py [hunk 1]
```diff
@@ -137,7 +130,20 @@ def default_to_pandas(self, pandas_op, *args, **kwargs):
...
     # treated differently.
 
-    lazy_execution = False
+    @property
+    def lazy_execution(self):
+        """
+        Whether underlying execution engine is designed to be executed in a lazy mode only.
+
+        If True, such QueryCompiler will be handled differently at the front-end in order
+        to reduce execution triggering as much as possible.
```
Comment (vnlitvinov) [inline modin/core/storage_formats/base/query_compiler.py:139]: This docstring is no longer true compared to what we're seeing in the implementation; plus I would rather see this change in PandasQueryCompiler, not BaseQueryCompiler (and leave this as a plain attribute in other places).

**Review by vnlitvinov — CHANGES_REQUESTED** [review]

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 2]
```diff
@@ -255,6 +255,21 @@ def default_to_pandas(self, pandas_op, *args, **kwargs):
         else:
             return result
 
+    @property
+    def lazy_execution(self):
+        """
+        Whether underlying execution engine is executed in a lazy mode.
+
+        If True, such QueryCompiler will be handled differently at the front-end in order
+        to reduce execution triggering as much as possible.
```
Comment (vnlitvinov) [inline modin/core/storage_formats/pandas/query_compiler.py:264]:
```suggestion
        Whether underlying Modin frame should be executed in a lazy mode.

        If True, such QueryCompiler will be handled differently at the front-end in order
        to reduce triggering the computation as much as possible.
```

**Review by vnlitvinov — COMMENTED** [review]
Good general idea, I'd like to fix the wording tho.

**Comment by anmyachev** [comment]
@vnlitvinov let's merge the PR

**Review by vnlitvinov — APPROVED** [review]

**Comment by anmyachev** [comment]
@mvashishtha could you review?

**Comment by mvashishtha** [comment]
@anmyachev it looks like the changes to `modin/pandas/base.py` are gone? I just see the PR adding `lazy_execution` but not using it anywhere.

**Comment by anmyachev** [comment]
> @anmyachev it looks like the changes to `modin/pandas/base.py` are gone? I just see the PR adding `lazy_execution` but not using it anywhere.

It's already used [here](https://github.com/modin-project/modin/blob/5991580cc8ac79c038d6dcde8b02d12feda06109/modin/pandas/base.py#L3174) and [here](https://github.com/modin-project/modin/blob/5991580cc8ac79c038d6dcde8b02d12feda06109/modin/pandas/base.py#L3419).

I just update `lazy_execution` implementation itself.

**Review by mvashishtha — APPROVED** [review]

**Pull request merged**